mm2src/db_common/src/sql_query.rs

```diff
@@ -0,0 +1,607 @@
+            .subquery()
+            .map_err(|e| SqlError::ToSqlConversionFailure(e.into()))?;
+        Ok(SqlQuery {
+            conn: subquery.0.conn,
+            sql_builder: SqlBuilder::select_from(subquery_sql),
+            params: subquery.0.params,
+            ordering: Vec::default(),
+        })
+    }
+
+    /// Add COUNT(field).
+    /// For more details see [`SqlBuilder::count`].
+    #[inline]
+    pub fn count<S: ToString>(&mut self, field: S) -> &mut Self {
+        self.sql_builder.count(field);
+        self
+    }
+
+    /// Add field.
+    /// For more details see [`SqlBuilder::field`].
+    #[inline]
+    pub fn field<S: ToString>(&mut self, field: S) -> &mut Self {
+        self.sql_builder.field(field);
+        self
+    }
+
+    /// Set OFFSET.
+    /// For more details see [`SqlBuilder::offset`].
+    #[inline]
+    pub fn offset<S: ToString>(&mut self, offset: S) -> &mut Self {
+        self.sql_builder.offset(offset);
+        self
+    }
+
+    /// Set LIMIT.
+    /// For more details see [`SqlBuilder::limit`].
+    #[inline]
+    pub fn limit<S: ToString>(&mut self, limit: S) -> &mut Self {
+        self.sql_builder.limit(limit);
+        self
+    }
+
+    /// Add GROUP BY part.
+    /// For more details see [`SqlBuilder::group_by`].
+    #[inline]
+    pub fn group_by<S: ToString>(&mut self, field: S) -> &mut Self {
+        self.sql_builder.group_by(field);
+        self
+    }
+
+    /// Add ORDER BY ASC.
+    /// For more details see [`SqlBuilder::order_asc`].
+    #[inline]
+    pub fn order_asc<S: ToString>(&mut self, field: S) -> &mut Self {
+        self.ordering.push(SqlOrdering::Asc(field.to_string()));
+        self
+    }
+
+    /// Add ORDER BY DESC.
+    /// For more details see [`SqlBuilder::order_desc`].
+    #[inline]
+    pub fn order_desc<S: ToString>(&mut self, field: S) -> &mut Self {
+        self.ordering.push(SqlOrdering::Desc(field.to_string()));
+        self
+    }
+
+    /// Join the given `table` and gives it the `alias`.
+    ///
+    /// Please note the function validates the given `table` name.
+    /// For more details see [`SqlBuilder::join`].
+    #[inline]
+    pub fn join_alias(&mut self, table: &str, alias: &'static str) -> SqlResult<&mut Self> {
+        validate_table_name(table)?;
+        validate_table_name(alias)?;
+        self.sql_builder.join(format!("{} {}", table, alias));
+        Ok(self)
+    }
+
+    /// Join constraint to the last JOIN part.
+    ///
+    /// Please note the function validates the given `table` name.
+    /// For more details see [`SqlBuilder::join`].
+    #[inline]
+    pub fn join(&mut self, table: &str) -> SqlResult<&mut Self> {
+        validate_table_name(table)?;
+        self.sql_builder.join(table);
+        Ok(self)
+    }
+
+    /// Join constraint to the last JOIN part [`SqlQuery::join`].
+    /// For more details see [`SqlBuilder::on_eq`].
+    #[inline]
+    pub fn on_eq<C1, C2>(&mut self, c1: C1, c2: C2) -> &mut Self
```

Comment:
I think `on_join_eq` will be a bit more clear name.

Reply:
Good idea! I think it's even better to combine `SqlQuery::join` and `SqlQuery::on_eq` methods into single `SqlQuery::join_on_eq`.

Reply:
I decided not to combine `SqlQuery::join` and `SqlQuery::on_eq`, but renamed `SqlQuery::on_eq` to `SqlQuery::on_join_eq` as you suggested.